RecoTracker/MeasurementDet/plugins/TkStackMeasurementDet.cc

```diff
@@ -20,16 +21,112 @@ void TkStackMeasurementDet::init(const MeasurementDet* lowerDet, const Measureme
 TkStackMeasurementDet::RecHitContainer TkStackMeasurementDet::recHits(const TrajectoryStateOnSurface& ts,
                                                                       const MeasurementTrackerEvent& data) const {
   RecHitContainer result;
-  /*
-  HitCollectorForRecHits collector( &fastGeomDet(), theMatcher, theCPE, result );
-  collectRecHits(ts, collector);
-*/
+
+  if (data.phase2OTVectorHits().empty())
+    return result;
+  LogTrace("MeasurementTracker") << " is not empty";
+  if (!isActive(data))
+    return result;
+  LogTrace("MeasurementTracker") << " and is active";
+
+  //find clusters to skip
+  const detset& lowerDetSet = data.phase2OTData().detSet(lowerDet()->index());
+  const detset& upperDetSet = data.phase2OTData().detSet(upperDet()->index());
+  std::vector<bool> skipClustersUpper(data.phase2OTClustersToSkip().empty() ? 0 : upperDetSet.size(), false);
+  std::vector<bool> skipClustersLower(data.phase2OTClustersToSkip().empty() ? 0 : lowerDetSet.size(), false);
+
+  const Phase2TrackerCluster1D* begin = nullptr;
+  if (!data.phase2OTData().handle()->data().empty()) {
+    begin = &(data.phase2OTData().handle()->data().front());
+  }
+  if (!data.phase2OTClustersToSkip().empty()) {
+    if (!lowerDetSet.empty()) {
+      for (const_iterator cil = lowerDetSet.begin(); cil != lowerDetSet.end(); ++cil) {
+        if (cil < begin) {
+          edm::LogError("IndexMisMatch") << "TkStackMeasurementDet cannot create hit because of index mismatch.";
+          return result;
+        }
+        unsigned int indexl = cil - begin;
+        if (data.phase2OTClustersToSkip()[indexl]) {
+          int iLocalL = std::distance(lowerDetSet.begin(), cil);
+          skipClustersLower[iLocalL] = true;
+        }
+      }
+    }
+    if (!upperDetSet.empty()) {
+      for (const_iterator ciu = upperDetSet.begin(); ciu != upperDetSet.end(); ++ciu) {
+        if (ciu < begin) {
+          edm::LogError("IndexMisMatch") << "TkStackMeasurementDet cannot create hit because of index mismatch.";
+          return result;
+        }
+        unsigned int indexu = ciu - begin;
+        if (data.phase2OTClustersToSkip()[indexu]) {
+          int iLocalU = std::distance(upperDetSet.begin(), ciu);
+          skipClustersUpper[iLocalU] = true;
+        }
+      }
+    }
+  }
+  DetId detIdStack = specificGeomDet().geographicalId();
+
+  auto iterator = data.phase2OTVectorHits().find(detIdStack);
+  if (iterator == data.phase2OTVectorHits().end())
+    return result;
+  for (const auto& vecHit : data.phase2OTVectorHits()[detIdStack]) {
+    if (!data.phase2OTClustersToSkip().empty()) {
+      if (skipClustersLower[vecHit.lowerCluster().key() - lowerDetSet.offset()])
+        continue;
+      if (skipClustersUpper[vecHit.upperCluster().key() - upperDetSet.offset()])
+        continue;
+    }
+    result.push_back(std::make_shared<VectorHit>(vecHit));
```

Comment:
would it make sense to make pseudo-shared pointer result here, e.g. `.push_back(std::shared_ptr<TrackingRecHit const>(&vectorHit, [](int *p) { }))`
unlike other tracker hits, VectorHits are pre-made.
@VinInn what do you think?